### PR TITLE
Add raw data and codeblock notebook  from paper

### LINKED ISSUE #775: Add code from the mitiq paper
The co-authors have agreed on making public the code relevant to the code snippets and plots appearing figures of the Mitiq paper (https://arxiv.org/abs/2009.04417). 

We discussed different possible routes. One was to have a separate repository, `mitiq-paper`. The one we are implementing, to ensure that the code keeps working and to reduce overhead, is to have the code snippets and plots included in the documentation of the library (run and tested with continuous integration).

We proposed to have the jupyter notebooks along with other examples in the **Mitiq Examples** section of the Users Guide in the documentation. 

We will also have accompanying data, that I propose we include simply in a `data/` subfolder there, under `docs/source/examples`. It is not heavy so it shouldn't be worth to selectively exclude it when cloning the repository. 

We also mentioned the fact that tags for the arXiv versions could help us pinpoint eventual modifications in the repository. 

I propose to also add some information about it under the **Research** section of the documentation. 

We did not create a Zenodo repository to avoid duplications of bibliographic records, with respect to the paper (once published) as this generates a DOI.

### PR DESCRIPTION
**Pull request by nathanshammah**
Close #775 adding notebooks and data from the Mitiq arXiv white paper. Please use #775 if you have comments on the way the code and data is being added, as a summary of previous discussions is contained there. 

There are currently a number of issues with the notebooks in this pull request: 

**Figure 3, previously Figure 2  in arxiv_v1**
- figure 2a and 2b work fine, although I get a repeated annoying warning message locally  
```
mitiq/mitiq/zne/inference.py:342: ExtrapolationWarning: You are pushing new data into a factory object despite its .reduce() 
method has already been called. Please make sure your intention is to append new data to the stack of previous data. 
Otherwise, the method .reset() can be used to clean the internal state of the factory.
```
**Figure 4, previously Figure 3 in arxiv_v1**
- figure 3 contains some errors and does not execute all blocks. I updated some names and imports so that they work fine. I fixed the easy ones, such as 
```
AttributeError: module 'mitiq' has no attribute 'execute_with_zne'
```
by changing the import of `execute_with_zne` with `mitiq.zne..` as it is not found at the top level;  or for `folding` that was put into `zne.scaling.folding`, etc.). Some errors are outstanding:
1. Some errors happen in `cirq`, and seem to be due to some code changes that happened in the library, for example this error
`TypeError: _resolve_parameters_() missing 1 required positional argument: 'recursive'`
2.  One type of error is raised in `mitiq`, 
<details>
`
---------------------------------------------------------------------------
UnfoldableCircuitError                    Traceback (most recent call last)
<ipython-input-17-421b13281678> in <module>
     15             return energy(circuit, radius_index=0, theta=theta, depo_noise_strength=p)
     16 
---> 17         zne = mitiq.zne.execute_with_zne(
     18             ansatz, executor, factory=pfac, scale_noise=mitiq.zne.scaling.fold_gates_at_random, num_to_average=1
     19         )

~/github/mitiq/mitiq/zne/zne.py in execute_with_zne(qp, executor, factory, scale_noise, num_to_average)
     60         raise ValueError("Argument `num_to_average` must be a positive int.")
     61 
---> 62     return factory.run(qp, executor, scale_noise, int(num_to_average)).reduce()
     63 
     64 

~/github/mitiq/mitiq/zne/inference.py in run(self, qp, executor, scale_noise, num_to_average)
    540 
    541         # Get all noise-scaled circuits to run
--> 542         to_run = self._generate_circuits(qp, scale_noise, num_to_average)
    543 
    544         # Get the list of keywords associated to each circuit in "to_run"

~/github/mitiq/mitiq/zne/inference.py in _generate_circuits(self, circuit, scale_noise, num_to_average)
    603         for scale_factor in self.get_scale_factors():
    604             for _ in range(num_to_average):
--> 605                 to_run.append(scale_noise(circuit, scale_factor))
    606         return to_run
    607 

~/github/mitiq/mitiq/interface/conversions.py in new_scaling_function(circuit, *args, **kwargs)
    221         circuit: QPROGRAM, *args: Any, **kwargs: Any
    222     ) -> QPROGRAM:
--> 223         scaled_circuit = atomic_converter(noise_scaling_function)(
    224             circuit, *args, **kwargs
    225         )

~/github/mitiq/mitiq/interface/conversions.py in qprogram_modifier(circuit, *args, **kwargs)
    170 
    171         # Modify the Cirq circuit.
--> 172         scaled_circuit = cirq_circuit_modifier(mitiq_circuit, *args, **kwargs)
    173 
    174         if kwargs.get("return_mitiq") is True:

~/github/mitiq/mitiq/zne/scaling/folding.py in fold_gates_at_random(circuit, scale_factor, seed, **kwargs)
    768     )
    769 
--> 770     return _apply_fold_mask(
    771         circuit,
    772         num_folds_mask,

~/github/mitiq/mitiq/zne/scaling/folding.py in _apply_fold_mask(circuit, num_folds_mask, squash_moments)
    508     Returns: The folded quantum circuit.
    509     """
--> 510     _check_foldable(circuit)
    511     circ_copy = deepcopy(circuit)
    512     measurements = _pop_measurements(circ_copy)

~/github/mitiq/mitiq/zne/scaling/folding.py in _check_foldable(circuit)
    102 
    103     if not has_unitary(circuit):
--> 104         raise UnfoldableCircuitError(
    105             "Circuit contains non-unitary channels which are not terminal "
    106             "measurements and cannot be folded."
`
</details>

```
UnfoldableCircuitError: Circuit contains non-unitary channels which are not terminal measurements and cannot be folded.
```
- For the Code Snippets file: The code snippets some time give errors, for example when pseudocode or not explicitly defined variables are used. I modified the pseudocode where it was easy in the first executor example (and actually I propose to change it also in the manuscript) and ask if folks know if there is a way to use `<variables>`, otherwise I guess we can just comment.  
- I propose to move, for consistency, the data loaded in one of the first blocks of the notebook of figure 3 into the data folder used for the others. 
- [ ] Some formatting of the notebooks should be made, even if we decided that they are not to be as illustrative as the other examples. We could also add something to the name, like `paper-figure3-vqe-h2` instead of `paper-figure3` and titles. They need to be included in the documentation via the examples' table of content. 
**Figure 5 **
- [ ] We need to add the PEC example yet.
While we mentioned this to be untangled from the arXiv updates, I suggest that we make sure that we the version of Mitiq pinned in the manuscript, all snippets work fine.

### CODE REVIEW

**Comment by rmlarose** [comment]
I like having notebook examples from the paper. I think it would be good to put these in a dedicated directory either in examples/ or parallel with examples/. However, this is blocked by #766 - none of these are being run, and the whole point of adding them is to make sure they are tested and actually work.

I'm very strongly against including data files. There are things like https://datadryad.org/stash/ for this.

**Comment by nathanshammah** [comment]
@rmlarose we discussed this at the Mitiq meeting, see the summary in #775. Basically here we are just adding the codeblocks from your jupyter notebook (which I slightly modified to avoid having to reinstall Qiskit and to have IBM Q secrets, commenting the last blocks). 

We also decided to adda the raw data supporting the plots as they are quite light.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb [hunk 1]
```diff
@@ -0,0 +1,1084 @@
...
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "wjnIQWX8d5UY"
+   },
+   "source": [
+    "# Mitiq paper codeblocks\n",
+    "\n",
+    "Code blocks from the main text of the Mitiq paper [1], updated to [arXiv v2](https://arxiv.org/abs/2009.04417)."
```
Comment (rmlarose) [inline docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:11]: I would not specify "v2" for future proofing. Just linking to arXiv is sufficient as this shows the latest version.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb [hunk 2]
```diff
@@ -0,0 +1,1084 @@
...
+    "outputId": "f2f56f66-9d37-4a17-f923-18c47a12ac5e"
+   },
+   "source": [
+    "```\n",
+    "!pip install qiskit==0.24.0 --quiet\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (rmlarose) [inline docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:988]: Why did you make the Qiskit cells markdown?

**Comment by rmlarose** [comment]
> (which I slightly modified to avoid having to reinstall Qiskit and to have IBM Q secrets, commenting the last blocks).

- What do you mean "reinstall Qiskit"? It hasn't been installed in the notebook prior to the `pip install qiskit` cell.
- You don't need IBMQ credentials with the `provider = qiskit.BasicAer` line.

There's a larger question of how to deal with imports / installs in documentation. Right now all the docs assume you have the environment set up so there are no pip installs. But I wrote this on Colab and needed `!pip install mitiq`. (I really like having documentation that you can run online, either colab or binder or other, and think we should move to this.)

Most importantly, we're still not testing documentation (#766). So for me adding the Mitiq paper codeblocks notebook is blocked by two things:

(1) Testing documentation (#766).
(2) How to deal with installs in documentation / should we make it easy to run documentation online?

**Review by crazy4pi314 — COMMENTED** [review]
LGTM

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 1 above)
Comment (nathanshammah) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb]: Ok i'll resolve that.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 2 above)
Comment (nathanshammah) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:988]: This is because I wanted to avoid issues with IBM logins.

**Comment by nathanshammah** [comment]
@rmlarose  Let's separate the issues. The goal of this PR is adding the codeblocks and supporting data for the paper. We can discuss testing and online interactive hosting of notebooks. 

Right now we are already making sure that all the code snippets in the examples in the documentation run correctly: If they didn't, the documentation would not build and an error would occur in CI, due to the way that the myst integration works, as it runs the notebooks each time and what is included in the html is the the output of that run. 

One can have `testcode` blocks also in the notebooks, as shown here (https://mitiq.readthedocs.io/en/stable/examples/test.html#testing) and here in the code 
https://github.com/unitaryfund/mitiq/blob/master/docs/source/examples/test.myst#L159-L173 that is not shown. See also my comment in #766 for more details.

I like the My Binder proposal. We wanted to do it and discussed it at the time the repo was not public (the [Mitiq alpha workshop](https://github.com/unitaryfund/mitiq/wiki/Mitiq-alpha-preview:-remote-workshop) used JupyterHub also as a way to avoid the issue that notebooks could not be run from a public repository), but then we all lost track of it. It should be quite straightforward.

**Comment by nathanshammah** [comment]
@rmlarose In the last commit I transformed the commented cells in the codeblocks from markdown to python code. I had it that way because locally I was getting an IBM Q error and I thought that was due to the fact that in the local environment I was I had no token for IBM Q identification and the error could arise also in CI.

File: docs/source/research.rst [hunk 3]
```diff
@@ -36,6 +36,22 @@ You can download the :download:`bibtex file <mitiq.bib>`.
 
 If you have developed new features for error mitigation, or found bugs in Mitiq, please consider `contributing <contributing.html>`_ your code.
```
Comment (crazy4pi314) [inline docs/source/research.rst:37]:
```suggestion
If you are considering implementing a new feature or bug fix for Mitiq, please check out some tips in `contributing <contributing.html>`_  to help make the contribution process easier.
```

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb [hunk 4]
```diff
@@ -0,0 +1,1087 @@
...
+    "id": "dyk7Hj1nPB48"
+   },
+   "outputs": [],
+   "source": [
+    "\n",
+    "import qiskit\n",
+    "\n",
+    "\n",
+    "# provider = qiskit.IBMQ.load_account()  # Use this with saved credentials.\n",
+    "provider = qiskit.BasicAer  # Workaround.\n",
```
Comment (crazy4pi314) [inline docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:1001]: I would just make this the default and just describe in a comment and just mention the local credentials (esp if mentioned in the paper)

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb [hunk 5]
```diff
@@ -0,0 +1,1087 @@
...
+    },
+    "id": "YTOaB_3tdymW",
+    "outputId": "38c2ce43-463f-4ab8-e402-fbc1d5877ce3"
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[K     |████████████████████████████████| 122kB 3.7MB/s \n",
```
Comment (crazy4pi314) [inline docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:38]: I take it there is no way to really suppress these dialogs here?

**Review by crazy4pi314 — COMMENTED** [review]
LGTM, just a couple of comments

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 5 above)
Comment (nathanshammah) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:38]: I don't know.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 4 above)
Comment (nathanshammah) [inline docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:1001]:
```suggestion
    "provider = qiskit.BasicAer  # Use of a simulator as backend.\n",
    "# provider = qiskit.IBMQ.load_account()  # Alternative way to run the blocks, with saved credentials.\n",
```

**Review by andreamari — APPROVED** [review]
LGTM. See the minor comment.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 4 above)
Comment (nathanshammah) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb]: This has now been changed to make the simulator just the default.

File: docs/source/research.rst [hunk 6]
```diff
@@ -34,7 +34,23 @@ If you are using Mitiq for your research, please cite the related `white paper <
...
+
+.. _code_data_mitiq_paper:
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Data and code supporting the Mitiq paper
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``mitiq-paper`` folder contains the data used in the ZNE plots of the Mitiq `white paper <https://arxiv.org/abs/2009.04417>`_.
+
+You can find the `code snippets <examples/mitiq-paper/mitiq-paper-codeblocks.html>`_ for the codeblocks present in the Mitiq paper in the `Examples` section of the documentation relative to the Mitiq paper section of the documentation.
```
Comment (nathanshammah) [inline docs/source/research.rst:48]:
```suggestion
You can find the `code snippets <examples/mitiq-paper/mitiq-paper-codeblocks.html>`_ for the codeblocks present in the Mitiq paper in the `Examples` section of the documentation relative to the Mitiq paper.
```

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 5 above)
Comment (rmlarose) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb:38]: It's from the pip install on colab. You can just delete the output cell.

File: docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb (hunk 4 above)
Comment (rmlarose) [reply docs/source/examples/mitiq-paper/mitiq-paper-codeblocks.ipynb]: Better way to do this (from ibmq example):

```python
if qiskit.IBMQ.stored_account():
    provider = qiskit.IBMQ.load_account()
    backend = provider.get_backend("ibmq_qasm_simulator")  # Set quantum computer here!
else:
    # Default to a simulator.
    backend = qiskit.BasicAer.backends()[0]
```

**Comment by nathanshammah** [comment]
This is ready for review, @rmlarose.

**Comment by rmlarose** [comment]
Superseded by #838.

**Pull request closed without merging**